cmd/cmd.go

```diff
@@ -862,16 +862,63 @@ func CopyHandler(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
+type pullFn func(ctx context.Context, name string, fn api.PullProgressFunc) error
+
+func getLocalPuller(insecure bool) (p pullFn, err error) {
+	p = func(ctx context.Context, name string, fn api.PullProgressFunc) error {
+		if err := initializeKeypair(); err != nil {
+			return err
+		}
+		opts := &server.RegistryOptions{
+			Insecure: insecure,
+		}
+		f := func(r api.ProgressResponse) {
+			fn(r)
+		}
+		return server.PullModel(ctx, name, opts, f)
+	}
+	return p, nil
+}
+
+func getAPIPuller(insecure bool) (p pullFn, err error) {
+	client, err := api.ClientFromEnvironment()
+	if err != nil {
+		return p, err
+	}
+	p = func(ctx context.Context, name string, fn api.PullProgressFunc) error {
+		if err := preflightCheck(ctx, client); err != nil {
+			return err
+		}
+		req := api.PullRequest{Name: name, Insecure: insecure}
+		return client.Pull(ctx, &req, fn)
+	}
+	return p, nil
+}
+
 func PullHandler(cmd *cobra.Command, args []string) error {
 	insecure, err := cmd.Flags().GetBool("insecure")
 	if err != nil {
 		return err
 	}
+	local, err := cmd.Flags().GetBool("local")
+	if err != nil {
+		return err
+	}
 
-	client, err := api.ClientFromEnvironment()
+	var pf pullFn
+	if local {
+		fmt.Println("pulling to local machine")
+		pf, err = getLocalPuller(insecure)
+	} else {
+		fmt.Println("requesting pull from server")
+		pf, err = getAPIPuller(insecure)
+	}
 	if err != nil {
 		return err
 	}
+	if pf == nil {
+		return fmt.Errorf("could not get puller: %w", err)
```

Comment:
Thanks!

Users should never see this error. The only time `pf` could be nil is if the code is refactored in future, and extra behaviours added. `pf` is currently assigned in the `if`/`else` statement, so there's no case when it could be nil.

However, in future, if someone refactored the code, they might change it so that `pf` is nil, in which case, they would get a panic. Rather than the future developer getting a panic, I figured it would be better for the developer to get an error message which localises the issue to this point of the codebase.

Reply:
However, this review has made me notice that the `err` should be nil at that point, so there's no need for the inclusion of wrapping an error.

I'll remove the nil check.